examples/nlp/sentence_embeddings_with_sbert.py

```diff
@@ -0,0 +1,463 @@
+"""
+# Sentence embeddings using siamese BERT-networks
+
+**Author:** [Mohammed Abu El-Nasr](https://github.com/abuelnasr0)<br>
+**Date created:** 2023/05/27<br>
+**Last modified:** 2023/05/27<br>
+**Description:** fine-tune a BERT or RoBERTa model to generate sentence embeddings using
+kerasNLP
+"""
+
+"""
+## Introduction
+"""
+
+"""
+BERT, and RoBERTa can be used for semantic textual similarity tasks, where two sentences
+are passed to model and the network predicts whether they are similiar or not. but what
+if we have a collection of large number of sentences, and we want to find the most
+similiar pairs in that collection, that will take n*(n-1)/2 inference computations, where
+n is the number of sentences in the collection. for example if n=10000, the required time
+will be 65 hours on a V100 GPU.<br>
+A common used method to overcome the time overhead issue, is to pass one sentence to the
+model, then average the output of the model, or take the first token (the [CLS] token)
+and use them as [sentence embedding](https://en.wikipedia.org/wiki/Sentence_embedding),
+then use a vector similarity measure like cosine similarity or Manhatten / Euclidean
+distance to find close sentences (semantically similair sentences). that will reduce the
+time to find the most similiar pairs in a collection of 10,000 sentences from 65 hours to
+5 seconds!<br>
+If we use BERT or RoBERa directly that will yield rather bad sentence embeddings. but if
+we fine-tune BERT using siamese network, that will generate simantically meaningful
+sentence embeddings. this will enable BERT to be used for new tasks. these tasks include:
+- Large-scale semantic similarity comparison
+- Clustring
+- Information retrival via semantic search
+<br>
+In this example, we will show how to fine-tune a BERT or RoBERTa model using siamese
+network so they will be able to produce semantically meaningful sentence embeddings and
+use them in semantic search and clustring example.<br>
+This method of fine-tuning was introduced is sentence-bert
+([SBERT](https://arxiv.org/abs/1908.10084))
+"""
+
+"""
+## Setup
+"""
+
+"""shell
+!pip install keras-nlp -q
+"""
+
+import keras_nlp
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow import keras
+import sklearn.cluster as cluster
+
+policy = keras.mixed_precision.Policy("mixed_float16")
+keras.mixed_precision.set_global_policy(policy)
+
+"""
+## Fine-tune the model using siamese networks
+"""
+
+"""
+[Siamese network](https://en.wikipedia.org/wiki/Siamese_neural_network) is a neural
+network architecture that contains two or more subnetworks. and the subnetworks share the
+same weights. it is used to generate feature vectors for each input and then compare them
+for simialrity.<br>
+For our example, the subnetworks will be RoBERTa model that have a pooling layer on top
```

Comment:
Correction: "the subnetworks will be `a` RoBERTa model that `has` a pooling layer"